src/main/kotlin/no/nav/familie/tilbake/iverksettvedtak/TilbakekrevingsvedtakBeregningService.kt

```diff
@@ -0,0 +1,225 @@
+import no.nav.familie.tilbake.common.isGreaterThanZero
+import no.nav.familie.tilbake.common.isLessThanZero
+import no.nav.familie.tilbake.common.isZero
+import no.nav.familie.tilbake.iverksettvedtak.domain.KodeResultat
+import no.nav.familie.tilbake.iverksettvedtak.domain.Tilbakekrevingsbeløp
+import no.nav.familie.tilbake.iverksettvedtak.domain.Tilbakekrevingsperiode
+import no.nav.familie.tilbake.kravgrunnlag.domain.Klassetype
+import no.nav.familie.tilbake.kravgrunnlag.domain.Kravgrunnlag431
+import no.nav.familie.tilbake.kravgrunnlag.domain.Kravgrunnlagsbeløp433
+import no.nav.familie.tilbake.kravgrunnlag.domain.Kravgrunnlagsperiode432
+import no.nav.familie.tilbake.vilkårsvurdering.domain.AnnenVurdering
+import org.springframework.stereotype.Service
+import java.math.BigDecimal
+import java.math.RoundingMode
+import java.util.UUID
+
+@Service
+class TilbakekrevingsvedtakBeregningService(private val tilbakekrevingsberegningService: TilbakekrevingsberegningService) {
+
+    fun beregnVedtaksperioder(behandlingId: UUID,
+                              kravgrunnlag431: Kravgrunnlag431): List<Tilbakekrevingsperiode> {
+        val beregningsresultat = tilbakekrevingsberegningService.beregn(behandlingId)
+
+        val kravgrunnlagsperioder = kravgrunnlag431.perioder.toList().sortedBy { it.periode.fom }
+        val beregnetPeioder = beregningsresultat.beregningsresultatsperioder.sortedBy { it.periode.fom }
+
+        // oppretter kravgrunnlagsperioderMedSkatt basert på månedligSkattebeløp
+        var kravgrunnlagsperioderMedSkatt = kravgrunnlagsperioder.associate { it.periode to it.månedligSkattebeløp }
+
+        val tilbakekrevingsperioder = mutableListOf<Tilbakekrevingsperiode>()
+        for (beregnetPeriode in beregnetPeioder) {
+            var perioder = lagTilbakekrevingsperioder(kravgrunnlagsperioder, beregnetPeriode)
+
+            // avrunding tilbakekrevesbeløp og uinnkrevd beløp
+            perioder = justerAvrunding(beregnetPeriode, perioder)
+
+            // skatt
+            // oppdaterer kravgrunnlagsperioderMedSkatt med gjenstående skatt
+            // (ved å trekke totalSkattBeløp fra månedligeSkattebeløp)
+            kravgrunnlagsperioderMedSkatt = oppdaterGjenståendeSkattetrekk(perioder, kravgrunnlagsperioderMedSkatt)
+            perioder = justerAvrundingSkatt(beregnetPeriode, perioder, kravgrunnlagsperioderMedSkatt.toMutableMap())
+
+            //renter
+            val totalTilbakekrevingsbeløp = beregnTotalTilbakekrevesbeløp(perioder)
+            var renteBeløp = BigDecimal.ZERO
+            if (beregnetPeriode.tilbakekrevingsbeløpUtenRenter != BigDecimal.ZERO) {
+                renteBeløp = beregnetPeriode.rentebeløp.multiply(totalTilbakekrevingsbeløp)
+                        .divide(beregnetPeriode.tilbakekrevingsbeløpUtenRenter, 0, RoundingMode.HALF_UP)
+            }
+            perioder.forEach { it.renter = renteBeløp }
+
+            tilbakekrevingsperioder.addAll(perioder)
+        }
```

Comment:
val tilbakekrevingsperioder =
                beregnetPeioder.map { beregnetPeriode ->
                    var perioder = lagTilbakekrevingsperioder(kravgrunnlagsperioder, beregnetPeriode)

                    // avrunding tilbakekrevesbeløp og uinnkrevd beløp
                    perioder = justerAvrunding(beregnetPeriode, perioder)

                    // skatt
                    // oppdaterer kravgrunnlagsperioderMedSkatt med gjenstående skatt
                    // (ved å trekke totalSkattBeløp fra månedligeSkattebeløp)
                    kravgrunnlagsperioderMedSkatt = oppdaterGjenståendeSkattetrekk(perioder, kravgrunnlagsperioderMedSkatt)
                    perioder = justerAvrundingSkatt(beregnetPeriode, perioder, kravgrunnlagsperioderMedSkatt)

                    //renter
                    val totalTilbakekrevingsbeløp = beregnTotalTilbakekrevesbeløp(perioder)
                    var renteBeløp = BigDecimal.ZERO
                    if (beregnetPeriode.tilbakekrevingsbeløpUtenRenter != BigDecimal.ZERO) {
                        renteBeløp = beregnetPeriode.rentebeløp.multiply(totalTilbakekrevingsbeløp)
                                .divide(beregnetPeriode.tilbakekrevingsbeløpUtenRenter, 0, RoundingMode.HALF_UP)
                    }
                    perioder.map { it.copy(renter = renteBeløp) }
                }.flatten()